### PR TITLE
Add crawler to test with a wide variety of websites and servers

### PR DESCRIPTION
**Pull request by sbstp**

### CODE REVIEW

File: crawler/Cargo.toml [hunk 1]
```diff
@@ -0,0 +1,12 @@
+[package]
+name = "crawler"
+version = "0.1.0"
+authors = ["Simon Bernier St-Pierre <git@sbstp.ca>"]
+edition = "2018"
+
+# See more keys and their definitions at https://doc.rust-lang.org/cargo/reference/manifest.html
```
Comment (adamreichold) [inline crawler/Cargo.toml]: This looks like a left over?

**Comment by adamreichold** [comment]
Do we want to make this an example so it at least it compiled with the tests? Or maybe create workspace? I fear that it might bitrot when it just lies around in the repository.

File: crawler/src/main.rs [hunk 2]
```diff
@@ -0,0 +1,72 @@
...
+    let mut listing: Listing = create_req("https://www.reddit.com/r/todayilearned/new.json")
+        .send()?
+        .json()?;
+
+    while let Some(after) = &listing.data.after {
+        if listing.data.children.is_empty() {
+            break;
+        }
+
+        process_items(&listing);
```
Comment (adamreichold) [inline crawler/src/main.rs]: Shouldn't we always process the items even if `listing.data.after` is `None`, e.g. for the last listing? So something like
```
let mut listing = create_req(...);

loop {
   process_items(&listing);

  if let Some(after) = listing.data.after {
    listing = create_req(...);
  } else {
    break;
  }
}
```

File: crawler/src/main.rs [hunk 3]
```diff
@@ -0,0 +1,72 @@
+use colored::*;
+use serde::{Deserialize, Serialize};
+
+type Listing = ProxyData<Posts>;
+
+#[derive(Debug, Deserialize, Serialize)]
+struct ProxyData<T> {
```
Comment (adamreichold) [inline crawler/src/main.rs]: Why is this generic instead of just containing `pub data: Posts`?

File: crawler/src/main.rs [hunk 4]
```diff
@@ -0,0 +1,72 @@
...
+        .header(
+            "User-Agent",
+            "Mozilla/5.0 (Windows NT 6.3; Win64; x64; rv:71.0) Gecko/20100101 Firefox/71.0",
+        )
+        .header("Accept", "*/*")
+}
+
+fn process_items(listing: &Listing) {
+    for item in &listing.data.children {
+        if !item.data.is_self {
```
Comment (adamreichold) [inline crawler/src/main.rs]: `if item.data.is_self { continue; }` could reduce rightward drift here.

File: crawler/src/main.rs [hunk 5]
```diff
@@ -0,0 +1,72 @@
...
+            match create_req(&item.data.url).send() {
+                Ok(resp) => {
```
Comment (adamreichold) [inline crawler/src/main.rs]: Maybe a guard like `Ok(resp) if resp.is_success() => println!(...)` would make this more uniform?

File: crawler/src/main.rs [hunk 6]
```diff
@@ -0,0 +1,72 @@
...
+struct Posts {
+    pub children: Vec<ProxyData<Post>>,
+    pub after: Option<String>,
+}
+
+#[derive(Debug, Deserialize, Serialize)]
+struct Post {
+    pub is_self: bool,
+    pub url: String,
+    pub subreddit: String,
```
Comment (adamreichold) [inline crawler/src/main.rs]: Why parse `subreddit` (and `ProxyData::kind`) if it isn't used?

File: crawler/Cargo.toml (hunk 1 above)
Comment (sbstp) [reply crawler/Cargo.toml]: This is created by `cargo new`

File: crawler/src/main.rs (hunk 3 above)
Comment (sbstp) [reply crawler/src/main.rs]: I use it at multiple places, in the Listing but also for the children.

File: crawler/src/main.rs (hunk 6 above)
Comment (sbstp) [reply crawler/src/main.rs]: No reason

File: crawler/src/main.rs (hunk 2 above)
Comment (sbstp) [reply crawler/src/main.rs]: Yes you're correct.

File: crawler/Cargo.toml (hunk 1 above)
Comment (adamreichold) [reply crawler/Cargo.toml]: But should we leave it here? I usually remove it first thing after creating a new crate.

File: crawler/src/main.rs (hunk 3 above)
Comment (adamreichold) [reply crawler/src/main.rs]: Ah, yes, I see it now.

**Pull request closed without merging**